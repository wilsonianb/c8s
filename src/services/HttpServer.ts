import * as Hapi from '@hapi/hapi'
import registerContainersController from '../controllers/containers'
import registerStaticController from '../controllers/static'
import registerProxyController from '../controllers/proxy'
import SelfTestCheck from '../services/SelfTestCheck'
import { Injector } from 'reduct'
import * as Inert from '@hapi/inert'
import * as Vision from '@hapi/vision'
import * as Handlebars from 'handlebars'
import * as path from 'path'
import Config from './Config'

import { create as createLogger } from '../common/log'
const log = createLogger('HttpServer')

export default class HttpServer {
  private config: Config
  private selfTestCheck: SelfTestCheck
  private server: Hapi.Server

  constructor (deps: Injector) {
    this.config = deps(Config)
    this.selfTestCheck = deps(SelfTestCheck)
    this.server = new Hapi.Server({
      uri: this.config.publicUri.replace(/\/+$/, ''),
      address: this.config.bindIp,
      port: this.config.port,
      mime: {
        override: {
          // Streaming responses shouldn't get buffered so for simplicity we'll
          // just turn off compression for them.
          'application/vnd.codius.raw-stream': {
            compressible: false
          }
        }
      }
    })

    registerContainersController(this.server, deps)
    registerStaticController(this.server, deps)
    registerProxyController(this.server, deps)
  }

  async init () {
    await this.server.register({ plugin: require('@hapi/h2o2') })
    await this.server.register(this.selfTestCheck.checkSelfTestPlugin)
    await this.server.register(Inert)
    await this.server.register(Vision)
    this.server.views({
      engines: {
        html: Handlebars
      },
      relativeTo: path.resolve(__dirname, '../'),
      path: 'public/templates'
    })

    this.server.route({
      method: 'GET',
      path: '/assets/app.bundle.js',
      handler: {
        file: path.join(__dirname, '../public/assets/app.bundle.js')
      }
    })

    this.server.route({
      method: 'GET',
      path: '/assets/proxy.bundle.js',
      handler: {
        file: path.join(__dirname, '../public/assets/proxy.bundle.js')
      }
    })
    await this.server.initialize()
  }

  async start () {
    await this.server.start()

    log.info('listening at %s', this.server.info.uri)
  }

  async stop () {
    await this.server.stop()

    log.info('server stopped')
  }

  getServer () {
    if (process.env.NODE_ENV === 'test') {
      return this.server
    }
    return null
  }
}
