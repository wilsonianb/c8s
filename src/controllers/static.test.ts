import HttpServer from '../services/HttpServer'
import reduct from 'reduct'
const deps = reduct()
const server = deps(HttpServer)

describe('static controller', () => {

  const options = {
    method: 'GET',
    url: '/'
  }

  beforeAll(async () => {
    await server.init()
  })

  afterAll(async () => {
    await server.stop()
  })

  test('responds with static index.html', async () => {
    const response = await server.getServer()!.inject(options)

    expect(response.statusCode).toBe(200)
    expect(response.result).toMatchSnapshot()
  })
})
