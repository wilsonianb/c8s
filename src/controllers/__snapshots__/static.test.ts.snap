// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static controller responds with static index.html 1`] = `
"<html>
  <head>
    <meta name='monetization' content=codius.org/local />
    <script>
      window.__INITIAL_PROPS__ = {\\"codiusVersion\\":\\"2.0.0\\"}
    </script>
  </head>
  <body>
    <div className=\\"App\\" id=\\"App\\">
      &lt;p data-reactroot&#x3D;&quot;&quot;&gt;Sorry! You need &lt;a href&#x3D;&quot;https://webmonetization.org&quot;&gt;Web Monetization&lt;/a&gt; to view this content.&lt;/p&gt;
    </div>
    <script src=/assets/app.bundle.js></script>
  </body>
</html>
"
`;
